rfcs/0025-reduce-scan-state-memory-usage.md

```diff
@@ -0,0 +1,97 @@
+## Summary
+[summary]: #summary
+
+This RFC analyzes current scan state memory usage and proposes a series of optimizations which can be peformed in order to allow us to meet our target TPS goals without consuming an astronomical amount of RAM.
+
+## Motivation
+[motivation]: #motivation
+
+In order to motivate these optimizations, we first need to analyze the expected memory usage of the scan state at our target TPS and slot time parameters. Below are the relevant calculations for computing the total memory usage (in bytes) of a single scan state, given 3 parameters of the scan state. `M` is the depth of each tree in the scan state, `T` is the total number of trees per scan state, and `D` is the scan state delay (we will discuss these in more detail later).
+
+![](https://latex.codecogs.com/gif.latex?NumberOfFullBranches%20%5Ctriangleq%20%5Csum_%7Bi%3D1%7D%5E%7B%5Cfrac%7BT-1%7D%7BD&plus;1%7D%7D%20%5Csum_%7Bj%3D0%7D%5E%7Bi%7D%202%5E%7BM-j%7D%20*%20%28%20D&plus;1%29)
+<!--
+```latex
+NumberOfFullBranches \triangleq \sum_{i=1}^{\frac{T-1}{D+1}} \sum_{j=0}^{i} 2^{M-j} * (D+1)
+```
+-->
+
+![](https://latex.codecogs.com/gif.latex?NumberOfEmptyBranches%20%5Ctriangleq%20NumberOfFullBranches%20-%20%28T-1%29*%282%5E%7BM&plus;1%7D-1%29)
+<!--
+```latex
+NumberOfEmptyBranches \triangleq NumberOfFullBranches - (T-1)*(2^{M+1}-1)
+```
+-->
+
+![](https://latex.codecogs.com/gif.latex?ScanState%20%5Ctriangleq%20%5C%5C%5Cindent%5Cindent%5Cindent%20%28T-1%29%20*%20%28%282*M-1%29%20*%20Word%29%20%5Cindent%5Cindent%5Cindent%5Cindent%5Cindent%5Cindent%5Cindent%7B%5Ccolor%7Bblue%7D%5Ctextbf%7Btree%20structure%20overhead%7D%7D%20%5C%5C%5Cindent%5Cindent%20&plus;%20NumberOfFullBranches%20*%20FullBranch%20%5C%5C%5Cindent%5Cindent%20&plus;%20NumberOfEmptyBranches%20*%20EmptyBranch%20%5C%5C%5Cindent%5Cindent%20&plus;%20%28T-1%29%20*%20%282%5EM%20*%20TreeLeaf%29%20%5C%5C%5Cindent%5Cindent%20&plus;%20%282%5E%7BM&plus;1%7D-1%20&plus;%20%28M-1%29%29%20*%20Word%20%5Cindent%5Cindent%5Cindent%5Cindent%5Cindent%5Cindent%7B%5Ccolor%7Bblue%7D%5Ctextbf%7Bempty%20tree%7D%7D)
+<!--
+```latex
+ScanState \triangleq
+\\\indent\indent\indent (T-1) * ((2*M-1) * Word)
+\indent\indent\indent\indent\indent\indent\indent{\color{blue}\textbf{tree structure overhead}}
+\\\indent\indent        + NumberOfFullBranches * FullBranch
+\\\indent\indent        + NumberOfEmptyBranches * EmptyBranch
+\\\indent\indent        + (T-1) * (2^M * TreeLeaf)
+\\\indent\indent        + (2^{M+1}-1 + (M-1)) * Word
+\indent\indent\indent\indent\indent\indent{\color{blue}\textbf{empty tree}}
+```
```

Comment:
This would then be `TreeStructureOverhead + (NumberOfFullBranches * FullBranch) + (NumberOfEmptyBranches * EmptyBranch) + (NumberOfFullLeaves * TreeLeaf) + (NumberOfEmptyLeaves * EmptyLeaf)` Where `EmptyLeaf = 2 * int + 1 * variant`